include/kernel.h

```diff
@@ -0,0 +1,534 @@
+ *     wakes up a thread (see #kevent). This is the most common scenario:
+ *     an event connected to an interrupt is fired, and a higher priority
+ *     thread waiting for the event is woken up.
+ * 
+ *   * Cooperatively: when a thread explicitly calls a blocking function
+ *     like #kthread_sleep or #kthread_wait_event to wait for an event.
+ *     The kernel will put the current thread to sleep and schedule the
+ *     highest-priority thread among the ready ones, including threads
+ *     that have a priority lower than the one that goes to sleep. When
+ *     the event is fired, the lower-priority thread will be preempted
+ *     to give back control to the thread that just woke up.
+ *
+ *   * Cooperatively: when a thread explicitly calls #kthread_yield to
+ *     leave the CPU to other threads that might be ready at the same time.
+ *     Notice that, given the hard real-time guarantee, this only makes
+ *     sense if there are multiple ready threads at the same priority level,
+ *     as calling #kthread_yield will not schedule a lower-priority thread.
+ *     It is normally not necessary to call #thread_yield.
+ *
+ * In addition to threads (#kthread_t) and events (#kevent_t), the kernel
+ * library offers a message passing communication primitive called "message
+ * box" (#kmbox_t). A mailbox is a circular buffer of messages called
+ * "mails"; any number of threads can try to send mails into a mailbox,
+ * and any number of threads can try to read mails from a mailbox. Once a
+ * thread receives a mail from a mailbox, the mail is removed and other
+ * threads will not see it anymore. 
+ *
+ * Mailboxes can be useful as communicative primitives among threads
+ * (eg: to send tasks to other threads), or a synchronization primitive
+ * to wait for multiple events (see #kmbox_attach_event).
+ *
+ * @{
+ */
+
+/**
+ * @brief a kernel thread for parallel execution
+ *
+ * This structure represents a thread that can be scheduled
+ * for parallel execution. Create a thread with #kthread_new,
+ * allocating the required memory for the stack.
+ *
+ * Normally, there's no need to explicitly manage the lifetime
+ * or state of a thread. Once the thread is created, it is
+ * immediately started, and the kernel will scheduled it when
+ * required. If the thread exits (by simply returning by its
+ * entry point function), it will be cleaned up and the memory
+ * released.
+ *
+ * #kthread_new returns a pointer to the thread (a pointer
+ * to kthread_t) that can be used to externally manage the
+ * thread like changing its priority or killing it.
+ */
+typedef struct kthread_s
+{
+	/** Pointer to the top of the stack, which contains the thread register state */
+	reg_block_t* stack_state;
+	/** Current state of interrupt depth for this thread */
+	int interrupt_depth;
+	/** Size of the stack in bytes */
+	int stack_size;
+	/** Name of thread (for debugging purposes) */
+	const char *name;
```

Comment:
What about keeping an id instead? I have seen in the examples that we even used userdata for that. Why not even remove this alltogether and provide a common userdata struct as a helper? This way we can kep this as minimal as possible and make it more extensible.

Reply:
I think that `name` is very useful for debugging. For instance, if the thread overflows its stack, you get a nice error screen saying that thread "foo" has overflown its stack of size N by at least X bytes. That is incredibly useful. 

At some point in the future, we might get the exception screen to also show the thread in which the exception was thrown, and we could even let the exception kill the thread but let the other threads continue running (libultra does this IIRC). 

I actually fail to see what we gain by removing the thread name.

Also I'm not sure about how this is related to `userdata`. `userdata` is just a generic argument to the user entry point, and it's actually a pretty common pattern in C for doing generic functors. This allows the caller to reuse the same function for multiple threads, passing some "context" to it (eg: the audio channel to play, or something).

Reply:
I haven't read the implementation yet and if it is integrated in such a way, it seems totally ok to have a name. I thought it was just a name the current thread can query for debugging on the consumer side, then it becomes equivalent to userdata.